### PR TITLE
Add Form base class `FormBaseClass`

### PR DESCRIPTION
**Pull request by rocky**
WIP - first cut of use BaseForms ...
    
and marking forms for going into $PrintForms and $OutputForms

<a href="https://gitpod.io/#https://github.com/Mathics3/mathics-core/pull/577"></a>

### CODE REVIEW

File: mathics/builtin/forms.py [hunk 1]
```diff
@@ -1,5 +1,11 @@
 """
-Print Forms
+Forms of Input and Output
+
+A <i>Form</i> format specifies the way Mathics Expression input is read or output written.
+
+A list of forms is contained in variable '$OutputForm'.
```
Comment (TiagoCavalcante) [inline mathics/builtin/forms.py:6]: Maybe "the variable contains..."?

File: mathics/builtin/forms.py [hunk 2]
```diff
@@ -37,224 +43,52 @@
...
-                for r in result
-            )
-        )
-
-
-class FullForm(Builtin):
+    # Using "__new__" is not optimal for what we want.
+    # We basically want to hook into class construction in order to
+    # detect certain class attributes so we can add them to a list.
+    # __new__ has this feature. However we dont' really need (or want)
```
Comment (TiagoCavalcante) [inline mathics/builtin/forms.py:66]: don't

File: mathics/builtin/forms.py [hunk 3]
```diff
@@ -37,224 +43,52 @@
...
+    # to do the memory allocation aspect that "__new__" is intended for.
+    # We considered __prepare__ and metaclass, instead bu could not figure
```
Comment (TiagoCavalcante) [inline mathics/builtin/forms.py:68]: bu -> but

File: mathics/builtin/forms.py [hunk 4]
```diff
@@ -351,59 +210,56 @@ class OutputForm(Builtin):
     summary_text = "plain-text output format"
 
 
-class MathMLForm(Builtin):
+class OutputForms_(Predefined):
+    r"""
+    <dl>
+      <dt>'$Output'
```
Comment (TiagoCavalcante) [inline mathics/builtin/forms.py:216]: $OutputForms

File: mathics/builtin/forms.py [hunk 5]
```diff
@@ -351,59 +210,56 @@ class OutputForm(Builtin):
...
      = ...
 
-    ## Test cases for Unicode - redo please as a real test
-    >> MathMLForm[\\[Mu]]
-    = ...
-
-    # This can causes the TeX to fail
-    # >> MathMLForm[Graphics[Text["\u03bc"]]]
-    #  = ...
+    Suppose now that we want to add a new format 'MyForm'. Initially, it does not belongs to '$PrintForms':
```
Comment (TiagoCavalcante) [inline mathics/builtin/forms.py:242]: ... does not belong...

**Review by TiagoCavalcante — APPROVED** [review]
LGTM - just some small things

File: mathics/builtin/forms.py (hunk 1 above)
Comment (rocky) [reply mathics/builtin/forms.py]: Thanks - all of these should be corrected now, but please double check.

**Comment by mmatera** [comment]
With this PR as it is, I found issues when I run the tests locally. In particular,

```
Testing section(s): TableForm
Testing section: Forms of Input and Output / TableForm
b'   1 ( 0): TEST TableForm[Array[a, {3,2}],TableDepth->1]'
result =!=wanted
----------------------------------------------------------------------
Test failed: TableForm in Reference of Built-in Symbols / Forms of Input and Output
Reference of Built-in Symbols
Result: $Aborted
Wanted: {a[1, 1], a[1, 2]}

{a[2, 1], a[2, 2]}

{a[3, 1], a[3, 2]}
Additional output:
$RecursionLimit::reclim: Recursion depth of 200 exceeded.

b'   2 ( 0): TEST Table[Style[Graphics[{EdgeForm[{Black}], RGBColor[r,g,b], Rectangle[]}], ImageSizeMultipliers->{0.2, 1}], {r,0,1,1/2}, {g,0,1,1/2}, {b,0,1,1/2}] // TableForm'
b'   3 ( 0): TEST TableForm[{}]'

```

and 

```
Testing section(s): MatrixForm
Testing section: Forms of Input and Output / MatrixForm
b'   1 ( 0): TEST Array[a,{4,3}]//MatrixForm'
b'   2 ( 0): TEST {{2*a, 0},{0,0}}//MatrixForm'
result =!=wanted
----------------------------------------------------------------------
Test failed: MatrixForm in Reference of Built-in Symbols / Forms of Input and Output
Reference of Built-in Symbols
Result: 2 ⁢ a   0

0       0

Wanted: 2 a   0

0     0

```

**Comment by mmatera** [comment]
Also,  here

https://github.com/Mathics3/mathics-core/blob/393186e562cae99d5306a65841569ad4700bfdef/mathics/builtin/forms.py#L299

here

https://github.com/Mathics3/mathics-core/blob/393186e562cae99d5306a65841569ad4700bfdef/mathics/builtin/forms.py#L331

and here

https://github.com/Mathics3/mathics-core/blob/393186e562cae99d5306a65841569ad4700bfdef/mathics/builtin/forms.py#L603

we have to change `apply_` by `eval_`

File: mathics/builtin/forms.py [hunk 6]
```diff
@@ -68,11 +68,6 @@ def __new__(cls, *args, **kwargs):
         """ """
         instance = super().__new__(cls, expression=False)
         name = cls.__name__
-        if name == "FormBaseClass":
```
Comment (rocky) [inline mathics/builtin/forms.py]: Why was this removed?
